Update Handling For When Linking New External Credentials Without IdP email and Refactor `def openid_connect`

aaronskiba opened the pull request:
Fixes #966

Changes proposed in this PR:
- Updates handling for when `if auth.info.email.nil? && user.nil?`
 - Previously, when this condition was true, the following code was executed:
     ```ruby
     flash[:notice] = _('Something went wrong, Please try signing up here.')
     redirect_to new_user_registration_path
     ```
  - Now, the executed code is contingent on whether or not the user is signed into the app (see `def handle_missing_email_for_new_sso_entry`).

- Keep user on /users/edit page after linking creds e8bfb5678256f7ef32fdc38243e48933381fc8f4
  - This redirect seems to make more sense than the prior redirect to `root_path`

- In addition, this pull request addresses the outstanding `Metrics/AbcSize` and `Metrics/MethodLength` rubocop offenses that existed within `def openid_connect`.

aaronskiba committed:
Improve flash msg handling for missing IdP email

aaronskiba committed:
Update comments in `def openid_connect`

aaronskiba committed:
Fix `openid_connect` rubocop offences

aaronskiba committed:
Make rubocop happy

aaronskiba committed:
Fix test for signed out user w/ missing IdP email

aaronskiba committed:
Add test for signed in user w/ missing IdP email

aaronskiba commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -152,5 +161,30 @@ def handle_new_sso_email_for_signed_out_user(auth, identifier_scheme)
+                                                    attrs: auth,
+                                                    identifiable: current_user)
+      flash[:notice] = _('Linked successfully')
+      redirect_to root_path
```
Should I change this too? Currently, after linking an account within Edit Profile, the user is then redirected to the root path. Would it be better to just keep them on the Edit Profile page?

aaronskiba replied:
I went ahead and made this change.

aaronskiba committed:
Update CHANGELOG.md

aaronskiba committed:
Keep user on `/users/edit` page after linking creds

- `def handle_new_sso_email_for_signed_in_user` is executed when a user links their credentials (for SSO purposes) on the `/users/edit` page.
- This change updates the `redirect_to` value so the user stays on the same page after the successful link.

aaronskiba committed:
Update test to match `redirect_to` change

- Updated the `redirect_to` value within this test to align with the update in the following commit: e8bfb5678256f7ef32fdc38243e48933381fc8f4

momo3404 commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -152,5 +161,30 @@ def handle_new_sso_email_for_signed_out_user(auth, identifier_scheme)
+        handle_new_sso_email_for_signed_in_user(auth, identifier_scheme)
+      # elsif user is signed in and trying to link via SSO, but the auth creds are already linked to another account
+      elsif user.id != current_user.id
+        handle_conflicting_sso_email_for_signed_in_user(identifier_scheme, user)
```
A comment here stating that `handle_conflicting_sso_email_for_signed_in_user` shows the appropriate error message would help in understanding the logic

momo3404 commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -152,5 +161,30 @@ def handle_new_sso_email_for_signed_out_user(auth, identifier_scheme)
+    end
+
+    def handle_new_sso_email_for_signed_in_user(auth, identifier_scheme)
+      current_user.identifiers << Identifier.create(identifier_scheme: identifier_scheme,
```
Is there a chance that there may be a duplicate `Identifier` in the db with the same credentials? I'm assuming not since the point of the function is to handle new SSO emails but I just wanted to double check.

aaronskiba replied:
I'll have to look into that one. I'm thinking `[Link your institutional credentials]` only renders when there is no existing Identifier for the signed in user, but it might be more complicated than that.

aaronskiba replied:
Also, just making a note of a conversation we had earlier.
`current_user.identifiers << Identifier.create(...)` may be redundant and `Identifier.create(...)` maybe all that is needed. I mentioned that our code is largely based off of `def handle_omniauth(scheme)`, and that action does not use `current_user.identifiers << `.

momo3404 commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -152,5 +161,30 @@ def handle_new_sso_email_for_signed_out_user(auth, identifier_scheme)
+    end
+
+    def handle_new_sso_email_for_signed_in_user(auth, identifier_scheme)
+      current_user.identifiers << Identifier.create(identifier_scheme: identifier_scheme,
```
Just curious - would it be better to have `create!` here to throw an error in case creating or saving the new value doesn't work?

aaronskiba replied:
That might be an improvement. Going back to this comment, https://github.com/portagenetwork/roadmap/pull/1003#issuecomment-2956735036, I can see that `def handle_omniauth(scheme)` also just uses `Identifier.create(`. Maybe I'll keep it as is for now and revisit this when we do finally use `handle_omniauth` rather than `openid_connect`.

momo3404 commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -152,5 +161,30 @@ def handle_new_sso_email_for_signed_out_user(auth, identifier_scheme)
       user
     end
+
+    def handle_openid_connect_for_signed_in_user(user, auth, identifier_scheme)
```
What do you think about renaming this one to `link_for_signed_in_user` ?

aaronskiba replied:
There's definitely some room for improvement. Because `def openid_connect` already  has `handle_openid_connect_for_signed_out_user(user, auth, identifier_scheme)`, I thought  `handle_openid_connect_for_signed_in_user(user, auth, identifier_scheme)` might be the most straightforward choice for now.

momo3404 replied:
I see, I think maintaining the consistency is more straightforward too.

momo3404 commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -152,5 +161,30 @@ def handle_new_sso_email_for_signed_out_user(auth, identifier_scheme)
+      flash[:notice] = _('Linked successfully')
+      redirect_to edit_user_registration_path
+    end
+
```
Do you think it would help to use `begin/rescue` here?

Something like 

```
begin
    current_user.identifiers << Identifier.create!(
      identifier_scheme: identifier_scheme,
      value: auth.uid,
      attrs: auth,
      identifiable: current_user
    )
    flash[:notice] = _('Linked successfully')
  rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotUnique => e
    flash[:alert] = _('There was a problem linking your account: %{error}') % { error: e.message }
  end

  redirect_to edit_user_registration_path
```

aaronskiba committed:
Update missing IdP email flash content

The https://cilogon.org/testidp/ behaviour seems to have changed with respect to required user attributes. These updated flash messages address that change.  (For more details, see https://github.com/portagenetwork/roadmap/issues/966#issuecomment-2937047435 )

aaronskiba committed:
Refactor to make rubocop happy

aaronskiba commented:
Hi @momo3404, I'll try to address some of these comments.

Overall, this is something I'd really like to better integrate with DMPRoadmap in the near future. `OmniauthCallbacksController#handle_omniauth` is very similar to `OmniauthCallbacksController#openid_connect` in terms of its underlying logic. There is also some additional code in DMPRoadmap's version of this controller (https://github.com/DMPRoadmap/roadmap/blob/main/app/controllers/users/omniauth_callbacks_controller.rb) that should allow us to eventually just replace the `openid_connect` action with `handle_omniauth` (specifically the code that is written to `"Dynamically build a handler for each omniauth provider"`)

momo3404 commented:
Testing note: Changing `if auth.info.email.nil? && user.nil?` to `if true` in `app/controllers/users/omniauth_callbacks_controller.rb` and trying to sing in via SSO is a good way to verify that the correct flash message is being rendered
![image](https://github.com/user-attachments/assets/eeac6743-dbda-480a-ba47-1e3a53cbbac1)
